### PR TITLE
[WIP] Fortran Parser for Sympy

### PR DESCRIPTION
**Pull request by Sc0rpi0n101**
#### References to other Issues or PRs

#16563

#### Brief description of what is fixed or changed
WIP PR for the Fortran parser for sympy
The parser can currently only parse basic assignments, basic arithmetic, function definitions.
.
#### TODO list

- [x] Make sure the external dependencies are optional
- [ ] Make sure the external dependencies are installed in Travis for the tests
- [x] Mark the module as provisional so that the API can be changed later if necessary
- [x] Make the module name all lowercase
- [x] Move the tests into a tests/ directory
- [x] Add a release notes entry

#### Release Notes

* parsing
  * Added a new Parser for parsing Fortran source code.

### CODE REVIEW

File: sympy/parsing/Fortran/fortran_parser.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+from lfortran.asr import asr
+from lfortran.semantic import kinds
+from lfortran.semantic.ast_to_asr import ast_to_asr
+from lfortran.ast import src_to_ast, print_tree
+from lfortran.asr.pprint import pprint_asr
+from sympy.parsing.sympy_parser import (parse_expr, standard_transformations, implicit_multiplication_application)
+from ast import *
```
Comment (czgdp1807) [inline sympy/parsing/Fortran/fortran_parser.py]: Is it possible to only import the specific classes/functions?

File: sympy/parsing/Fortran/fortran_parser.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        for item in node.items:
+            self.visit(item)
+
+    def visit_Assignment(self, node):
+        #TODO: Arithmatic Assignment
+        if isinstance(node.target, asr.Variable):
+            target = node.target
+            value = node.value
+            if (value.type == asr.Variable):
+                new_node = Assign(targets = [Name(id = target.name, ctx = Store())], value = Name(id = value.name, ctx = Store()))
```
Comment (czgdp1807) [inline sympy/parsing/Fortran/fortran_parser.py]: Please split this line into two

File: sympy/parsing/Fortran/fortran_parser.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                if isinstance(value.op, asr.Add):
+                    bin_op = Add()
+                elif isinstance(value.op, asr.Sub):
+                    bin_op = Sub()
+                elif isinstance(value.op, asr.Div):
+                    bin_op = Div()
+                elif isinstance(value.op, asr.Mul):
+                    bin_op = Mult()
+
+                new_node = Assign(targets = [Name(id = target.name, ctx = Store())], value = BinOp(left = Name(id = value.left.name, ctx = Load()), op = bin_op, right = Name(id = value.right.name, ctz = Load())))
```
Comment (czgdp1807) [inline sympy/parsing/Fortran/fortran_parser.py]: > Please split this line into two

+1

File: sympy/parsing/Fortran/fortran_parser.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            elif isinstance(op, asr.Div):
+                new_node = Expr(value = BinOp(left = Name(id = lhs.name, ctx = Load()),
+                 op =Div(), right = Name(id = rhs.name, ctx = Load())))
+            elif isinstance(op, asr.Add):
+                new_node = Expr(value = BinOp(left = Name(id = lhs.name, ctx = Load()),
+                 op =Add(), right = Name(id = rhs.name, ctx = Load())))
+            elif isinstance(op, asr.Sub):
+                new_node = Expr(value = BinOp(left = Name(id = lhs.name, ctx = Load()),
+                 op =Sub(), right = Name(id = rhs.name, ctx = Load())))
+            else:
```
Comment (czgdp1807) [inline sympy/parsing/Fortran/fortran_parser.py]: `else` not needed.

File: sympy/parsing/Fortran/fortran_parser.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        for arg_iter in node.args:
+            fn_args.append(arg(arg = arg_iter.name, annotation=None))
+        for i in node.body:
+            fn_ast = call_visitor(i)
+        fn_body = fn_ast.body
+        #for sym in node.symtab.symbols:
+        #    fn_body.append(Expr(value = Name(id = node.symtab.symbols[sym].name, ctx = Store())))
+        fn_return = Name(id = node.return_var.name, ctx = Load())
+        fn_body.append(Return(value = Name(id = fn_return, ctx = Store())))
+        new_node = FunctionDef(name = fn_name, args = arguments(args = fn_args,vararg = None, kwarg =None, defaults = [], kw_defaults = []), decorator_list = [], body = fn_body)
```
Comment (czgdp1807) [inline sympy/parsing/Fortran/fortran_parser.py]: > Please split this line into two.

+1

File: sympy/parsing/Fortran/fortran_parser.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+def test_assignment():
+    src2= """\
+    integer :: a, b, c, d, e, f
+    c = a + b
+    d = a - b
+    e = a * b
+    f = a / b
+    """
+    res2 = """\
```
Comment (czgdp1807) [inline sympy/parsing/Fortran/fortran_parser.py]: Is it possible to use ("<string-1>", "<string-2>") for concatenation?

File: sympy/parsing/Fortran/fortran_parser.py (hunk 1 above)
Comment (Sc0rpi0n101) [reply sympy/parsing/Fortran/fortran_parser.py]: Yes, It is also compulsory, Travis doesn't allow `import *`
but it is a work in progress. I will import specific functions later

File: sympy/parsing/Fortran/fortran_parser.py (hunk 2 above)
Comment (Sc0rpi0n101) [reply sympy/parsing/Fortran/fortran_parser.py]: Yes, I will

File: sympy/parsing/Fortran/fortran_parser.py (hunk 3 above)
Comment (Sc0rpi0n101) [reply sympy/parsing/Fortran/fortran_parser.py]: Sure

File: sympy/parsing/Fortran/fortran_parser.py (hunk 4 above)
Comment (czgdp1807) [reply sympy/parsing/Fortran/fortran_parser.py]: Ah! sorry its needed. Please resolve this conversation.

**Comment by asmeurer** [comment]
Use lowercase names for the file names. Uppercase names can cause issues on case insensitive filesystems.

File: sympy/parsing/Fortran/fortran_parser.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+t_src = """\
+integer function f(a, b) result(r)
+integer, intent(in) :: a, b
+r = a + b
+end function
+"""
+src_to_sympy(t_src)
+
+
+def test_function():
```
Comment (asmeurer) [inline sympy/parsing/Fortran/fortran_parser.py]: Please put all tests in a separate test_fortran_parser.py file in a tests/ subdirectory

**Pull request closed without merging**